File: apps/claude-of-duty/src/engine.rs
//! The frame loop and the shared context handed to every subsystem.
//!
//! Ported from `C:/dev/Claude-of-Duty/src/core/engine.js:1-158`, minus the
//! Three.js scene/camera construction (`engine.js:28-35`, plus the camera aspect
//! updates in `resize`) — those are the renderer's business and land with the
//! render arm of the port, not with the loop.
//!
//! The Engine owns the frame loop and the shared context handed to every
//! subsystem. It does NOT know what any subsystem does — it only sequences them.
//!
//! Frame order (`engine.js:11-17`, unchanged):
//!   1. `input.begin_frame()`
//!   2. `fixed_update(FIXED_DT)` xN   — physics, deterministic gameplay
//!   3. `update(dt)`                  — animation, cameras, AI decisions
//!   4. `late_update(dt)`             — anything that must observe final transforms
//!   5. render subsystem draws
//!   6. `input.end_frame()`
//!
//! Steps 1 and 6 are marked in [`Engine::step`] and do nothing yet:
//! `core/input.js` is not ported, and inventing a stand-in for it here would put
//! input policy in the loop that is supposed to be blind to it.
//!
//! ## What the browser arm keeps
//!
//! `start()`, `stop()` and the `requestAnimationFrame` trampoline
//! (`engine.js:100-116`) are not here. They are the platform edge — a browser
//! frame source — and the loop they drive is [`Engine::step`], which the source
//! already exposes for exactly this reason ("so the capture harness can pump
//! frames by hand"). The wasm bootstrap calls `step`; so does every test; so does
//! a headless capture. One loop, several clocks.
//!
//! The root seed is likewise explicit. The source picks
//! `config.deterministic ? 0x5eed1234 : Math.random()` — the only `Math.random()`
//! in the entire game. A constructor argument makes every run reproducible and
//! puts the choice of entropy at the platform edge where it belongs.

use std::cell::{Cell, RefCell};

use axiom_kernel::Seconds;

use crate::config::{Config, FIXED_DT, FIXED_STEP, MAX_SUBSTEPS};
use crate::error::CoreError;
use crate::events::{DispatchFailure, EventBus};
use crate::registry::{Phase, Registry, SystemRef, Subsystem};
use crate::rng::Rng;

/// The source's capture seed (`engine.js:26`) — what `config.deterministic`
/// selects so a screenshot is byte-identical run to run.
pub const CAPTURE_SEED: u32 = 0x5eed_1234;

/// The id the source reserves for the drawing subsystem (`engine.js:145`).
pub const RENDER_SYSTEM_ID: &str = "render";

/// The engine clock, as the source's `this.time` object.
///
/// The scalars are `f64` because the accumulator must not drift and because a JS
/// number *is* an `f64` — narrowing here would change the stepping. The kernel
/// `Seconds` accessors are the boundary: a subsystem is handed dimensioned time,
/// never a naked float.
#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Time {
    /// Seconds since start, scaled.
    pub elapsed: f64,
    /// Unscaled wall-clock seconds since start.
    pub raw: f64,
    /// Last frame delta, scaled and clamped.
    pub dt: f64,
    /// Fixed step.
    pub fixed: f64,
    /// Interpolation alpha between the last two physics steps, 0..1.
    pub alpha: f64,
    pub scale: f64,
    pub frame: u64,
}

impl Time {
    fn start() -> Self {
        Time {
            fixed: FIXED_DT,
            scale: 1.0,
            ..Time::default()
        }
    }

    /// The frame delta as a dimensioned duration.
    pub fn dt_seconds(self) -> Seconds {
        Seconds::finite_or_zero(self.dt as f32)
    }

    /// Time since start as a dimensioned duration.
    pub fn elapsed_seconds(self) -> Seconds {
        Seconds::finite_or_zero(self.elapsed as f32)
    }
}

/// The shared context every subsystem is stepped with — the source's `ctx`
/// object, minus the Three.js handles.
///
/// It is a borrowed view rather than a stored struct: the engine hands out one
/// per phase, so the fields it mutates between phases (`time.alpha`, notably)
/// cannot be observed half-updated.
pub struct Ctx<'a> {
    pub config: &'a Config,
    pub events: &'a EventBus,
    pub time: &'a Time,
    /// The root random stream. Every subsystem takes `ctx.rng.borrow_mut()
    /// .fork()` once at init and never touches this again — that discipline is
    /// what keeps one system's edits from reshuffling another's sequence.
    pub rng: &'a RefCell<Rng>,
    registry: &'a Registry,
}

impl Ctx<'_> {
    /// `ctx.get(id)` — throwing lookup.
    pub fn get(&self, id: &str) -> Result<SystemRef, CoreError> {
        self.registry.get(id)
    }

    /// `ctx.peek(id)` — non-throwing lookup.
    pub fn peek(&self, id: &str) -> Option<SystemRef> {
        self.registry.peek(id)
    }

    /// `ctx.has(id)`.
    pub fn has(&self, id: &str) -> bool {
        self.registry.has(id)
    }
}

/// The frame loop.
pub struct Engine {
    pub config: Config,
    registry: Registry,
    events: EventBus,
    rng: RefCell<Rng>,
    time: Time,
    /// The fixed-step accumulator.
    ///
    /// A `Cell` because the fixed loop drains it *while* a [`Ctx`] borrowing the
    /// registry is alive; JS has no such constraint, and splitting the drain out
    /// of the loop would change when `time.alpha` becomes visible to a
    /// `fixed_update`.
    accum: Cell<f64>,
    /// Timestamp of the previous [`Engine::step`], in milliseconds — the units
    /// `performance.now()` reports in, kept so the clamp below reads the same.
    last: f64,
}

impl Engine {
    /// Build an engine. `root_seed` is the seed of the one random stream every
    /// subsystem forks from; pass [`CAPTURE_SEED`] for a reproducible capture.
    pub fn new(config: Config, root_seed: u32) -> Self {
        Engine {
            config,
            registry: Registry::new(),
            events: EventBus::new(),
            rng: RefCell::new(Rng::new(root_seed)),
            time: Time::start(),
            accum: Cell::new(0.0),
            last: 0.0,
        }
    }

    /// `engine.add(SystemClass, opts)`. Returns the shared handle rather than
    /// `this`: `Result` does not chain, and the handle is what a caller needs to
    /// reach a system by its concrete type later.
    pub fn add(&mut self, system: impl Subsystem + 'static) -> Result<SystemRef, CoreError> {
        self.registry.add(system)
    }

    pub fn registry(&self) -> &Registry {
        &self.registry
    }

    pub fn events(&self) -> &EventBus {
        &self.events
    }

    pub fn time(&self) -> Time {
        self.time
    }

    /// Set the simulation time scale — slow-motion, hit-stop, pause. The only
    /// field of `time` the source ever writes from outside the loop.
    pub fn set_time_scale(&mut self, scale: f64) {
        self.time.scale = scale;
    }

    /// Resolve the dependency order and initialise every system in it.
    ///
    /// The source is `async` (it awaits asset loads) and logs any system whose
    /// init exceeds 50ms. Neither survives: there is nothing to await yet, and
    /// per-system timing is telemetry that belongs behind the kernel's telemetry
    /// sink rather than a `console.info` in the loop.
    pub fn init(&mut self) -> Result<(), CoreError> {
        let order = self.registry.resolve()?;
        let ctx = Ctx {
            config: &self.config,
            events: &self.events,
            time: &self.time,
            rng: &self.rng,
            registry: &self.registry,
        };
        for system in order {
            system.borrow_mut().init(&ctx)?;
        }
        Ok(())
    }

    /// Viewport changed. The source reads the size off the canvas; here the
    /// platform edge that owns the canvas passes it in.
    pub fn resize(&self, width: u32, height: u32) -> Result<Vec<DispatchFailure>, CoreError> {
        let width = width.max(1);
        let height = height.max(1);
        let systems = self.registry.with(Phase::Resize)?;
        let ctx = Ctx {
            config: &self.config,
            events: &self.events,
            time: &self.time,
            rng: &self.rng,
            registry: &self.registry,
        };
        for system in systems {
            system.borrow_mut().resize(width, height, &ctx);
        }
        Ok(self.events.emit("resize", &(width, height)))
    }

    /// Re-anchor the frame clock, so the next [`Engine::step`] measures a delta
    /// from `now` rather than from zero. The source does this in `start()`
    /// (`this._last = performance.now()`); without it the first frame after a
    /// long pause eats the 0.1s clamp.
    pub fn reset_clock(&mut self, now: f64) {
        self.last = now;
    }

    /// Advance one frame. `now` is a monotonic timestamp in **milliseconds**,
    /// the unit `performance.now()` reports in.
    pub fn step(&mut self, now: f64) -> Result<(), CoreError> {
        // Clamp so a tab-switch or a breakpoint doesn't teleport the simulation.
        let raw_dt = 0.1f64.min(0.0f64.max((now - self.last) / 1000.0));
        self.last = now;
        self.time.raw += raw_dt;
        self.time.dt = raw_dt * self.time.scale;
        self.time.elapsed += self.time.dt;
        self.time.frame += 1;

        // 1. input.begin_frame() — `core/input.js` is not ported yet.

        let fixed_systems = self.registry.with(Phase::FixedUpdate)?;
        let update_systems = self.registry.with(Phase::Update)?;
        let late_systems = self.registry.with(Phase::LateUpdate)?;
        let render_system = self
            .registry
            .peek(RENDER_SYSTEM_ID)
            .filter(|s| s.borrow().phases().contains(&Phase::Render));

        // 2. The fixed loop. Scoped so `time.alpha` can be written the instant
        //    it ends — a `fixed_update` sees the *previous* frame's alpha, which
        //    is what the source does and what an interpolating physics system
        //    reads.
        let steps = {
            let ctx = Ctx {
                config: &self.config,
                events: &self.events,
                time: &self.time,
                rng: &self.rng,
                registry: &self.registry,
            };
            self.accum.set(self.accum.get() + self.time.dt);
            let mut steps = 0u32;
            while self.accum.get() >= FIXED_DT && steps < MAX_SUBSTEPS {
                for system in &fixed_systems {
                    system.borrow_mut().fixed_update(FIXED_STEP, &ctx);
                }
                self.accum.set(self.accum.get() - FIXED_DT);
                steps += 1;
            }
            steps
        };
        if steps == MAX_SUBSTEPS {
            // Shed the backlog rather than spiral: a frame that could not keep
            // up drops the un-simulated time instead of handing the next frame
            // an even bigger debt.
            self.accum.set(0.0);
        }
        self.time.alpha = self.accum.get() / FIXED_DT;

        {
            let ctx = Ctx {
                config: &self.config,
                events: &self.events,
                time: &self.time,
                rng: &self.rng,
                registry: &self.registry,
            };
            let dt = self.time.dt_seconds();
            // 3. update
            for system in &update_systems {
                system.borrow_mut().update(dt, &ctx);
            }
            // 4. late_update
            for system in &late_systems {
                system.borrow_mut().late_update(dt, &ctx);
            }
            // 5. the render subsystem draws
            if let Some(system) = render_system {
                system.borrow_mut().render(&ctx);
            }
        }

        // 6. input.end_frame() — `core/input.js` is not ported yet.

        Ok(())
    }

    /// Tear down: dispose every system in reverse dependency order, then drop
    /// every event subscription.
    pub fn dispose(&mut self) -> Result<(), CoreError> {
        let mut order = self.registry.ordered()?;
        order.reverse();
        for system in order {
            system.borrow_mut().dispose();
        }
        self.events.clear();
        Ok(())
    }
}

File: apps/shmup/src/core/engine.js
import * as THREE from 'three';
import { Registry, EventBus } from './registry.js';
import { boot } from './profile.js';
import { Streamer } from './streaming.js';
import { FIXED_DT, MAX_SUBSTEPS } from './config.js';
import { Input } from './input.js';
import { Rng } from './rng.js';

/**
 * The Engine owns the frame loop and the shared context handed to every
 * subsystem. It does NOT know what any subsystem does — it only sequences them.
 *
 * Frame order:
 *   1. input.beginFrame()
 *   2. fixedUpdate(FIXED_DT) xN   — physics, deterministic gameplay
 *   3. update(dt)                 — animation, cameras, AI decisions
 *   4. lateUpdate(dt)             — anything that must observe final transforms
 *   5. render subsystem draws
 *   6. input.endFrame()
 */
export class Engine {
  /**
   * @param opts.bakery  optional worker pool for pure procedural work, handed
   *   in by the composition root. Core does not build it, because the recipes
   *   live in subsystems and core does not import subsystems; the Engine only
   *   publishes it on ctx so any subsystem can use it during init.
   */
  constructor({ canvas, config, bakery = null }) {
    this.canvas = canvas;
    this.config = config;
    this.bakery = bakery;
    this.registry = new Registry();
    this.events = new EventBus();
    this.input = new Input(canvas, config);
    this.rng = new Rng(config.deterministic ? 0x5eed1234 : (Math.random() * 2 ** 32) >>> 0);
    this._traceForks();

    this.scene = new THREE.Scene();
    this.camera = new THREE.PerspectiveCamera(config.fov, 1, 0.05, 1200);
    this.camera.rotation.order = 'YXZ';

    /** Separate scene+camera for the first-person viewmodel, drawn with its own
     *  near plane so hands/weapon never clip into world geometry. */
    this.viewScene = new THREE.Scene();
    this.viewCamera = new THREE.PerspectiveCamera(60, 1, 0.005, 12);

    this.time = {
      /** Seconds since start, scaled. */ elapsed: 0,
      /** Unscaled wall-clock seconds since start. */ raw: 0,
      /** Last frame delta, scaled and clamped. */ dt: 0,
      /** Fixed step. */ fixed: FIXED_DT,
      /** Interpolation alpha between the last two physics steps, 0..1. */ alpha: 0,
      scale: 1,
      frame: 0,
    };

    this.ctx = {
      engine: this,
      scene: this.scene,
      camera: this.camera,
      viewScene: this.viewScene,
      viewCamera: this.viewCamera,
      canvas,
      config,
      events: this.events,
      input: this.input,
      time: this.time,
      rng: this.rng,
      bakery,
      get: (id) => this.registry.get(id),
      peek: (id) => this.registry.peek(id),
      has: (id) => this.registry.has(id),
    };

    /** Deferred construction, drained across frames once the loop is live. */
    this.streamer = new Streamer({ budgetMs: config.streamBudgetMs ?? 6 });

    this._accum = 0;
    this._last = 0;
    this._running = false;
    this._onResize = () => this.resize();
  }

  /**
   * `?rngtrace=1` — record a stack for every fork of the ROOT stream.
   *
   * Every procedural thing in this game hangs off this one stream, and each
   * subsystem's seed is decided purely by HOW MANY forks preceded it. So a
   * change that adds, removes or moves a single root fork reseeds everything
   * downstream of it and repaints the game — which the pixel gate reports as
   * "all eleven shots changed", true and useless.
   *
   * This turns that into one line. It found the real case it was written for:
   * `Viewmodel` forked ctx.rng from inside its constructor, so it sat at
   * position 6 only because weapons happened to init before fx — and hoisting
   * the other forks into prepare() silently moved it to position 10, reseeding
   * fx, ai, ui and audio. Read it with `node tools/rngprobe.mjs`.
   *
   * Off unless asked for: it allocates an Error per fork.
   */
  _traceForks() {
    const on = typeof location !== 'undefined' &&
      new URLSearchParams(location.search).get('rngtrace') === '1';
    if (!on) return;
    const trace = [];
    const orig = this.rng.fork.bind(this.rng);
    this.rng.fork = () => {
      trace.push(String(new Error().stack).split(/\r?\n/).slice(2, 5).join(' | '));
      return orig();
    };
    this.rngForkTrace = trace;
  }

  add(SystemClass, opts) {
    this.registry.add(new SystemClass(opts));
    return this;
  }

  async init() {
    const order = this.registry.resolve();

    // PREPARE PASS — claim seeds and start pure precomputation, before any
    // subsystem blocks the main thread.
    //
    // Two things happen here and nothing else may:
    //   1. every subsystem forks its private Rng off ctx.rng;
    //   2. a subsystem with pure, seed-driven precomputation (the procedural
    //      texture bakes) queues it on the bakery and keeps the promise.
    //
    // WHY IT IS A SEPARATE PASS. Those bakes are ~3.5 s of value-noise
    // evaluation, and if each one is queued inside its own subsystem's init
    // there is nothing left to overlap it with — `ai` is second-to-last, so it
    // awaits ~2.4 s with the main thread idle. Started here, every bake is in
    // flight on a worker while `render`, `world` and `weapons` build their
    // object graphs on this thread. Measured with tools/bootprofile.mjs.
    //
    // WHY THE FORK MOVED WITH IT. The fork is a single `u32()` draw from
    // ctx.rng, and its POSITION in that stream decides every seed downstream of
    // it. Hoisting only the baking subsystems would shift everyone else's draw
    // and repaint the whole game. This pass runs in `resolve()` order — the same
    // order the init loop below uses — so the draw sequence is byte-for-byte
    // what it was when the forks lived in init(). The pixel gate is the proof.
    boot.time('engine.prepare', () => {
      for (const sys of order) sys.prepare?.(this.ctx);
    });

    // Sequential BY CONTRACT, not by accident: `resolve()` returns dependency
    // order and a subsystem may read any of its declared deps out of ctx during
    // its own init. Overlapping two of these is only legal between subsystems
    // with no path between them in the dep graph — and, because almost every
    // expensive init here is main-thread GPU work on one shared context, the
    // legal pairs would interleave rather than overlap. Profile before assuming
    // concurrency is the win: `node tools/bootprofile.mjs`.
    for (const sys of order) {
      const id = sys.constructor.id;
      const t0 = performance.now();
      await boot.timeAsync(`init:${id}`, () => sys.init?.(this.ctx));
      const ms = performance.now() - t0;
      if (ms > 50) console.info(`[engine] ${id} init ${ms.toFixed(0)}ms`);
    }
    // Collect the deferred half of every subsystem's construction, in the same
    // dependency order the inits ran in — so the RNG draws and the object-graph
    // mutations happen in exactly the sequence they did when this work was
    // inline. See src/core/streaming.js.
    boot.time('engine.collectStream', () => {
      for (const sys of order) this.streamer.add(sys.constructor.id, sys.stream?.(this.ctx));
    });

    boot.time('engine.attach', () => {
      this.input.attach();
      // Keys pressed from here on are BUFFERED even though no frame has run yet
      // (see Input.beginFrame), so this is where the game stops discarding the
      // player's input — distinct from, and earlier than, when it acts on it.
      boot.milestone('input-armed');
      addEventListener('resize', this._onResize);
      this.resize();
    });
    return this;
  }

  /**
   * Finish every subsystem's deferred construction NOW, rather than across
   * frames. The capture harness calls this before it raises `__READY__`: a shot
   * of a half-streamed world is not a regression, it is a different picture, and
   * the pixel gate cannot tell the difference. Also the right call for any
   * caller that wants "fully loaded" to mean fully loaded.
   */
  drainStream() {
    return boot.timeAsync('engine.drainStream', () => this.streamer.drainAll());
  }

  resize() {
    const w = Math.max(1, this.canvas.clientWidth || innerWidth);
    const h = Math.max(1, this.canvas.clientHeight || innerHeight);
    this.camera.aspect = w / h;
    this.camera.updateProjectionMatrix();
    this.viewCamera.aspect = w / h;
    this.viewCamera.updateProjectionMatrix();
    for (const sys of this.registry.with('resize')) sys.resize(w, h, this.ctx);
    this.events.emit('resize', { width: w, height: h });
  }

  start() {
    if (this._running) return;
    this._running = true;
    this._last = performance.now();
    this._loop = this._loop.bind(this);
    requestAnimationFrame(this._loop);
  }

  stop() {
    this._running = false;
  }

  _loop(now) {
    if (!this._running) return;
    requestAnimationFrame(this._loop);
    this.step(now);
  }

  /** Advance one frame. Exposed so the capture harness can pump frames by hand. */
  step(now = performance.now()) {
    const t = this.time;
    // Clamp so a tab-switch or a breakpoint doesn't teleport the simulation.
    const rawDt = Math.min(0.1, Math.max(0, (now - this._last) / 1000));
    this._last = now;
    t.raw += rawDt;
    t.dt = rawDt * t.scale;
    t.elapsed += t.dt;
    t.frame++;

    this.input.beginFrame();

    this._accum += t.dt;
    let steps = 0;
    const fixedSystems = this.registry.with('fixedUpdate');
    while (this._accum >= FIXED_DT && steps < MAX_SUBSTEPS) {
      for (const sys of fixedSystems) sys.fixedUpdate(FIXED_DT, this.ctx);
      this._accum -= FIXED_DT;
      steps++;
    }
    if (steps === MAX_SUBSTEPS) this._accum = 0; // shed backlog rather than spiral
    t.alpha = this._accum / FIXED_DT;

    for (const sys of this.registry.with('update')) sys.update(t.dt, this.ctx);
    for (const sys of this.registry.with('lateUpdate')) sys.lateUpdate(t.dt, this.ctx);

    const renderSystem = this.registry.peek('render');
    if (typeof renderSystem?.render === 'function') renderSystem.render(this.ctx);

    // THE number progressive boot exists to move: when the player first sees
    // the game, as opposed to when the last deferred thing finishes loading.
    boot.milestone('first-frame');

    this.input.endFrame();

    // AFTER the frame is drawn, never before: the first frame must pay nothing
    // for streaming, and every frame after it should present what it has before
    // spending anything on what it does not have yet.
    const streaming = this.streamer.step();
    if (!streaming && !this._streamDone) {
      this._streamDone = true;
      this.events.emit('stream:done', this.streamer.stats);
    }
  }

  dispose() {
    this.stop();
    removeEventListener('resize', this._onResize);
    this.input.detach();
    for (const sys of [...this.registry.ordered].reverse()) sys.dispose?.();
    this.bakery?.dispose();
    this.events.clear();
  }
}
